src/main/java/apoc/export/cypher/formatter/CreateCypherFormatter.java

```diff
@@ -13,37 +23,280 @@
+
+
+        Function<Relationship, Map<String, Object>> keyMapper = (rel) -> {
+            try (Transaction tx = db.beginTx()) {
+                Node start = rel.getStartNode();
+                String startLabels = CypherFormatterUtils.formatAllLabels(start, uniqueConstraints, indexNames);
+
+                // define the end labels
+                Node end = rel.getEndNode();
+                String endLabels = CypherFormatterUtils.formatAllLabels(end, uniqueConstraints, indexNames);
+
+                // define the type
+                String type = rel.getType().name();
+
+                // create the path
+                Map<String, Object> key = new HashMap<>();
+                key.put("start", new AbstractMap.SimpleImmutableEntry<>(startLabels, CypherFormatterUtils.getNodeIdProperties(start, uniqueConstraints).keySet()));
+                key.put("type", type);
+                key.put("end", new AbstractMap.SimpleImmutableEntry<>(endLabels, CypherFormatterUtils.getNodeIdProperties(end, uniqueConstraints).keySet()));
+                tx.success();
+                return key;
+            }
+        };
+        Map<Map<String, Object>, List<Relationship>> groupedData = StreamSupport.stream(relationship.spliterator(), true)
+                .collect(Collectors.groupingByConcurrent(keyMapper));
+
+        AtomicInteger propertiesCount = new AtomicInteger(0);
+        AtomicInteger batchCount = new AtomicInteger(0);
+
+        int unwindBatchSize = exportConfig.getUnwindBatchSize();
+
+        String start = "start";
+        String end = "end";
+        groupedData.forEach((path, relationshipList) -> {
+            int relSize = relationshipList.size();
+            relCount.addAndGet(relSize);
+            for (int i = 0; i < relSize; i++) {
+                writeUnwindStart(exportConfig, out, batchCount);
+                batchCount.incrementAndGet();
+                Relationship rel = relationshipList.get(i);
+
+                Map<String, Object> props = rel.getAllProperties();
+                // start element
+                out.append("{");
+
+                // start node
+                Node startNode = rel.getStartNode();
+                writeNodeIds(uniqueConstraints, out, start, startNode);
+
+                out.append(", ");
+
+                // end node
+                Node endNode = rel.getEndNode();
+                writeNodeIds(uniqueConstraints, out, end, endNode);
+
+                // properties
+                out.append(", ");
+                out.append("properties: ");
+                out.append("{");
+                if (!props.isEmpty()) {
+                    out.append(CypherFormatterUtils.formatProperties("", props, true).substring(2));
+                    propertiesCount.addAndGet(props.size());
+                }
+                out.append("}");
+
+                // end element
+                out.append("}");
+
+                boolean isEnd = i == relSize - 1;
```

Comment:
this is quite some computing :)
definitely prone to off-by one errors :)

Reply:
ok